mm/tempesta_mm.c

```diff
@@ -22,166 +22,71 @@
-#define PGNUM			(1 << pgorder)
-#define PGNUM4K			(PGNUM * (1 << HUGETLB_PAGE_ORDER))
+#define MAX_MEMSZ		65536 * HPAGE_SIZE /* 128GB per node */
+#define MIN_MEMSZ		16 * HPAGE_SIZE	/* 32MB per node */
+#define DEFAULT_MEMSZ		256 * HPAGE_SIZE /* 512MB */
 
-static int pgorder = DEFAULT_PGORDER;
-static gfp_t gfp_f = GFP_HIGHUSER | __GFP_COMP | __GFP_THISNODE | __GFP_ZERO
-		     | __GFP_RETRY_MAYFAIL;
+static unsigned long dbmem = DEFAULT_MEMSZ;
 static TempestaMapping map[MAX_NUMNODES];
-/*
- * Modern x86-64 has not more than 512GB RAM per physical node.
- * This is very large amount of memory, but it will be freed when
- * initialization phase ends.
- */
-static struct page *greedy[GREEDY_ARNUM] __initdata = { 0 };
 
-static int __init
-tempesta_setup_pages(char *str)
+static unsigned long
+__dbsize_mb(unsigned long size)
 {
-	get_option(&str, &pgorder);
-	if (pgorder < MIN_PGORDER) {
-		pr_err("Tempesta: bad dbmem value %d, must be [%d:%d]\n",
-		       pgorder, MIN_PGORDER, MAX_PGORDER);
-		pgorder = MIN_PGORDER;
-	}
-	if (pgorder > MAX_PGORDER) {
-		pr_err("Tempesta: bad dbmem value %d, must be [%d:%d]\n",
-		       pgorder, MIN_PGORDER, MAX_PGORDER);
-		pgorder = MAX_PGORDER;
-	}
+	if (size >= SZ_1M)
+		return size / (1024 * 1024);
 
-	return 1;
+	return 0;
 }
-__setup("tempesta_dbmem=", tempesta_setup_pages);
 
-/**
- * The code is somewhat stollen from mm/hugetlb.c.
- */
-static struct page *
-tempesta_alloc_hpage(int nid)
+static int __init
+tempesta_setup_pages(char *str)
 {
-	struct page *p;
-
-	p = alloc_pages_node(nid, gfp_f, HUGETLB_PAGE_ORDER);
-	if (!p)
-		return NULL;
-
-	count_vm_event(HTLB_BUDDY_PGALLOC);
+	unsigned long min = __dbsize_mb(MIN_MEMSZ) * nr_online_nodes;
+	unsigned long max = __dbsize_mb(MAX_MEMSZ) * nr_online_nodes;
+	unsigned long raw_dbmem = memparse(str, NULL);
 
-	__ClearPageReserved(p);
+	/* Count memory per node */
+	dbmem = round_up(raw_dbmem / nr_online_nodes, HPAGE_SIZE);
 
-	return p;
-}
+	if (dbmem < MIN_MEMSZ) {
+		pr_err("Tempesta: bad dbmem value %lu(%luM), must be [%luM:%luM]"
+		       "\n", raw_dbmem, __dbsize_mb(raw_dbmem), min, max);
+		dbmem = MIN_MEMSZ;
+	}
+	if (dbmem > MAX_MEMSZ) {
+		pr_err("Tempesta: bad dbmem value %lu(%luM), must be [%luM:%luM]"
+		       "\n", raw_dbmem, __dbsize_mb(raw_dbmem), min, max);
+		dbmem = MAX_MEMSZ;
+	}
 
-static void
-tempesta_free_hpage(struct page *p)
-{
-	__free_pages(p, HUGETLB_PAGE_ORDER);
+	return 1;
 }
+__setup("tempesta_dbmem=", tempesta_setup_pages);
 
-/**
- * Greedely alloc huge pages and try to find continous region organized
- * by sorted set of allocated pages. When the region is found, all pages
- * out of it are returned to system.
+/*
+ * Something like @prep_compound_gigantic_page() for Hugetlb and
+ * @prep_compound_page(). The main difference we don't set the desctructor.
  */
-static struct page *
-tempesta_alloc_contmem(int nid)
+static void prep_huge_compound_page(struct page *page, unsigned int order)
```

Comment:
The only difference between this function and `[prep_compound_gigantic_page](https://elixir.bootlin.com/linux/v5.10.35/C/ident/prep_compound_gigantic_page)` is that we don't call `__ClearPagereserved`. Why we don't use `prep_compound_gigantic_page`?

Reply:
`prep_compound_gigantic_page`. However, `mem_map_next` looks unnecessary here I will change calc of p to simple: `p = page + i`.